cmd/clusterctl/client/cluster/topology.go

```diff
@@ -0,0 +1,641 @@
+			ctx,
+			client.ObjectKey{Namespace: cluster.GetNamespace(), Name: cluster.GetName()},
+			storedCluster,
+		); err != nil {
+			if apierrors.IsNotFound(err) {
+				// The Cluster does not exist in the server. Nothing more to do here.
+				continue
+			}
+			return errors.Wrapf(err, "failed to get Cluster %s/%s", cluster.GetNamespace(), cluster.GetName())
+		}
+		originalJSON, err := json.Marshal(storedCluster)
+		if err != nil {
+			return errors.Wrapf(err, "failed to convert Cluster %s/%s to json", storedCluster.Namespace, storedCluster.Name)
+		}
+		modifiedJSON, err := json.Marshal(cluster)
+		if err != nil {
+			return errors.Wrapf(err, "failed to convert Cluster %s/%s", cluster.GetNamespace(), cluster.GetName())
+		}
+		// Apply the modified object to the original one, merging the values of both;
+		// in case of conflicts, values from the modified object are preserved.
+		originalWithModifiedJSON, err := jsonpatch.MergePatch(originalJSON, modifiedJSON)
+		if err != nil {
+			return errors.Wrap(err, "failed to apply modified json to original json")
+		}
+		if err := json.Unmarshal(originalWithModifiedJSON, &cluster.Object); err != nil {
+			return errors.Wrap(err, "failed to convert modified json to object")
+		}
+	}
+	return nil
+}
+
+// defaultAndValidate runs the defaulting and validation webhooks on the
+// ClusterClass and Cluster objects in the input thus replicating the real kube-apiserver flow
+// when applied.
+// Nb. Perform ValidateUpdate only if the object is already in the cluster. In all other cases,
+// ValidateCreate is performed.
+// *Important Note*: We cannot perform defaulting and validation on provider objects as we have access to
+// that code.
+func (t *topologyClient) defaultAndValidate(ctx context.Context, in *DryRunInput, apiReader client.Reader) error {
+	// Enable the ClusterTopology feature gate so that the defaulter and validators do not complain.
+	// Note: We don't need to disable it later because the CLI is short lived.
+	if err := feature.Gates.(featuregate.MutableFeatureGate).Set(fmt.Sprintf("%s=%v", feature.ClusterTopology, true)); err != nil {
+		return errors.Wrapf(err, "failed to enable %s feature gate", feature.ClusterTopology)
+	}
+
+	// From the inputs gather all the objects that are not Clusters or ClusterClasses.
+	// These objects will be used when initializing a dryrun client to use in the webhooks.
+	filteredObjs := filterObjects(
+		in.Objs,
+		clusterv1.GroupVersion.WithKind("ClusterClass"),
+		clusterv1.GroupVersion.WithKind("Cluster"),
+	)
+	objs := []client.Object{}
+	for _, o := range filteredObjs {
+		objs = append(objs, o)
+	}
+	// Creating a dryrun client will a fall back to the apiReader client (client to the underlying Kubernetes cluster)
+	// allows the defaulting and validation webhooks to complete actions to could potentially depend on other objects in the cluster.
+	// Example: Validation of cluster objects will use the client to read ClusterClasses.
```

Comment:
This should work with the fakeclient too - right ? Assuming it's correctly created etc.

Reply:
yes.